ADD: 웨이브 이동 포탈 추가

Jo-sm94 opened the pull request:
ShooterBaseGameMode
- 웨이브 클리어 시 웨이브전환에서 포탈 생성으로 변경

TeleportPortal
- 포탈 영역 내에 5초 이상 머무를 시 다음 웨이브 레벨 로드
- 포탈 생성 위치 확인 필요

[Content/PlayerCharacter/GamePlayAbility/GA_WeaponReload.uasset]파일은 블루프린트 노드 확인하다가 위치 수정된게 같이 올라가 버렸습니다....
정원님이랑 얘기해서 파일 백업하고 정원님이 수정할 부분이 있어서 그 부분이랑 같이 올려주신다고 하셨습니다.

Jo-sm94 committed:
Merge branch 'Develop' into feature/GameMode

Jo-sm94 committed:
Merge branch 'Develop' into feature/GameMode

Jo-sm94 committed:
Merge branch 'Develop' into feature/GameMode

Jo-sm94 committed:
ADD: 웨이브 이동 포탈 추가

ShooterBaseGameMode
- 웨이브 클리어 시 웨이브전환에서 포탈 생성으로 변경

TeleportPortal
- 포탈 영역 내에 5초 이상 머무를 시 다음 웨이브 레벨 로드
- 포탈 생성 위치 확인 필요

Jo-sm94 committed:
ADD: 웨이브 이동 포탈 추가

ShooterBaseGameMode
- 웨이브 클리어 시 웨이브전환에서 포탈 생성으로 변경

TeleportPortal
- 포탈 영역 내에 5초 이상 머무를 시 다음 웨이브 레벨 로드
- 포탈 생성 위치 확인 필요

Jo-sm94 committed:
Merge branch 'feature/GameMode' of https://github.com/NbcampUnreal/3rd_4th-Team8-CH3-Project into feature/GameMode

hwarang98 commented on Source/Shooter/Private/GameModes/TeleportPortal.cpp:
```diff
@@ -0,0 +1,99 @@
+	//UI 카운트다운 시간 초 전달 부분 예정
+	UE_LOG(LogTemp, Warning, TEXT("CountDown : %d"), CountTime);
+
+	if (CountTime <= 1) {
```
해당코드는 counttime이 1일때도 타이머가 종료되는걸로 보입니다! 혹시 의도된걸까요?

Jo-sm94 replied:
실제 포탈에 오버랩되면 다음 레벨로 이동되기까지의 시간을 UI상에서 보여줄꺼같은데 그때 5 4 3 2 1 실제 이동 이부분이 0으로 계산하면 5 4 3 2 1 0 실제 이동 이렇게 되버려서 1로 설정했습니다
같은 이유로 CountTime 도 6으로 설정

hwarang98 commented on Source/Shooter/Private/GameModes/TeleportPortal.cpp:
```diff
@@ -0,0 +1,99 @@
+void ATeleportPortal::OnOverlapEnd(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex)
+{
+
+	if (OtherActor == OverlapActor)
```
`if (OverlapActor && OtherActor == OverlapActor)` 이렇게 비교하는게 비교적 더 안전합니다!
왜냐하면 혹시나 OverlapActor가 이전 로직에서 해제된 상태에서 이 함수가 호출되더라도 방어할 수 있습니다.

Jo-sm94 replied:
그렇네요 OverlapBegin에서는 OverlapActor도 확인했는데 OnOverlapEnd에서는 빼먹었습니다.
수정하겠습니다

hwarang98 commented on Source/Shooter/Private/GameModes/TeleportPortal.cpp:
```diff
@@ -0,0 +1,99 @@
+
+	if (OtherActor == OverlapActor)
+	{
+		GetWorldTimerManager().ClearTimer(PortalTimerHandle);
```
로직상에는 문제없어보이는 훌륭한 코드입니다!
다만 한가지 제안 드리고싶은건 해당 if 문 안의 로직도 함수화 시키는게 유지보수성에서 더 좋다고 판단됩니다!
===예제===
```cpp
void ATeleportPortal::ResetPortalState()
{
	GetWorldTimerManager().ClearTimer(PortalTimerHandle);
	OverlapActor = nullptr;
	CountTime = 6;
	UE_LOG(LogTemp, Warning, TEXT("Player left portal zone. Timer cancelled."));
}
```
이렇게 함수를 따로빼고 if문 로직안에서는 `ResetPortalState()` 함수만 실행시키면됩니다! (함수의 재사용)

Jo-sm94 committed:
ADD: 웨이브 이동 포탈 추가 (피드백 수정)

ShooterBaseGameMode
 - 웨이브 클리어 시 웨이브전환에서 포탈 생성으로 변경

TeleportPortal
 - 포탈 영역 내에 5초 이상 머무를 시 다음 웨이브 레벨 로드
